beetsplug/tidal.py

```diff
@@ -0,0 +1,1050 @@
+        if item.album:
+            self._log.debug("Searching using album")
+            for album in self._search_album(item.album):
+                self._log.debug(
+                    f"Using album {album.name} from {album.artist.name}"
+                )
+                for track in album.tracks():
+                    tracks.append(track)
+                    trackids.append(track.id)
+
+        # Reverse list so the more specific result is first
+        tracks = reversed(tracks)
+
+        # Remove duplicates
+        trackids = []
+        newtracks = []
+        for track in tracks:
+            if track.id in trackids:
+                self._log.debug(f"Removing duplicate track {track.id}")
+                continue
+            else:
+                trackids.append(track.id)
+                newtracks.append(track)
+
+        tracks = newtracks
+        return tracks
+
+    def _search_album(self, query, limit=10, offset=0):
+        """Searches TIDAL for albums matching the query
+
+        :param query: The search string to use
+        :type query: str
+        :param limit: Maximum number of items to return, defaults to 10
+        :type limit: int, optional
+        :param offset: Offset the items to retrieve, defaults to 0
+        :type offset: int, optional
+        :return: A list of tidalapi Albums
+        :rtype: list
+        """
+
+        self._log.debug(f"_search_album query {query}")
+        results = self._tidal_search(query, [tidalapi.Album], limit, offset)
+
+        candidates = results["albums"]
+
+        # isinstance call is required here as the top_hit can be any tidalapi type
+        if results["top_hit"] and isinstance(
+            results["top_hit"], tidalapi.album.Album
+        ):
+            candidates.insert(0, results["top_hit"])
+
+        self._log.debug(f"_search_album found {len(candidates)} results")
+        return candidates
+
+    def _search_track(self, query, limit=10, offset=0):
+        """Searches TIDAL for tracks matching the query
+
+        :param query: The search string to use
+        :type query: str
+        :param limit: Maximum number of items to return, defaults to 10
+        :type limit: int, optional
+        :param offset: Offset the items to retrieve, defaults to 0
+        :type offset: int, optional
+        :return: A list of tidalapi Tracks
+        :rtype: list
+        """
+        self._log.debug(f"_search_track raw query {query}")
+
+        results = self._tidal_search(query, [tidalapi.Track], limit, offset)
+        candidates = results["tracks"]
+
+        # isinstance call is required here as the top_hit can be any tidalapi type
+        if results["top_hit"] and isinstance(
+            results["top_hit"], tidalapi.media.Track
+        ):
+            candidates.insert(0, results["top_hit"])
+
+        self._log.debug(f"_search_track found {len(candidates)} results")
+        return candidates
```

Comment:
There's a lot in common between these two methods. Could this potentially be abstracted?